homeassistant/components/valve/__init__.py

```diff
@@ -0,0 +1,278 @@
+
+    @property
+    def reports_position(self) -> bool:
+        """Return True if entity reports position, False otherwise."""
+        if hasattr(self, "_attr_reports_position"):
+            return self._attr_reports_position
+        if hasattr(self, "entity_description"):
+            return self.entity_description.reports_position
+        raise ValueError(f"'reports_position' not set for {self.entity_id}.")
+
+    @property
+    def current_valve_position(self) -> int | None:
+        """Return current position of valve.
+
+        None is unknown, 0 is closed, 100 is fully open.
+        """
+        return self._attr_current_valve_position
+
+    @property
+    def device_class(self) -> ValveDeviceClass | None:
+        """Return the class of this entity."""
+        if hasattr(self, "_attr_device_class"):
+            return self._attr_device_class
+        if hasattr(self, "entity_description"):
+            return self.entity_description.device_class
+        return None
+
+    @property
+    @final
+    def state(self) -> str | None:
+        """Return the state of the valve."""
+        reports_position = self.reports_position
+        if self.is_opening:
+            self._valve_is_last_toggle_direction_open = True
+            return STATE_OPENING
+        if self.is_closing:
+            self._valve_is_last_toggle_direction_open = False
+            return STATE_CLOSING
+        if reports_position is True:
+            if (current_valve_position := self.current_valve_position) is None:
+                return None
+            position_zero = current_valve_position == 0
+            return STATE_CLOSED if position_zero else STATE_OPEN
+        if (closed := self.is_closed) is None:
+            return None
+        return STATE_CLOSED if closed else STATE_OPEN
+
+    @final
+    @property
+    def state_attributes(self) -> dict[str, Any]:
+        """Return the state attributes."""
+
+        return {ATTR_CURRENT_POSITION: self.current_valve_position}
+
+    @property
+    def supported_features(self) -> ValveEntityFeature:
+        """Flag supported features."""
+        return self._attr_supported_features
+
+    @property
+    def is_opening(self) -> bool | None:
+        """Return if the valve is opening or not."""
+        return self._attr_is_opening
+
+    @property
+    def is_closing(self) -> bool | None:
+        """Return if the valve is closing or not."""
+        return self._attr_is_closing
+
+    @property
+    def is_closed(self) -> bool | None:
+        """Return if the valve is closed or not."""
+        return self._attr_is_closed
+
+    def open_valve(self) -> None:
+        """Open the valve."""
+        if not ValveEntityFeature.SET_POSITION | self.supported_features:
+            raise NotImplementedError()
+        self.set_valve_position(100)
+
+    async def async_open_valve(self) -> None:
+        """Open the valve."""
+        await self.hass.async_add_executor_job(ft.partial(self.open_valve))
```

Comment:
This means a call to the open or close services will always require an executor job.

Instead of this, we should add make the `SERVICE_OPEN_VALVE` and `SERVICE_CLOSE_VALVE` call intermediate functions `_async_handle_open_valve` + `_async_handle_close_valve` or similar. These functions will then directly call `async_set_valve_position` for valves which support setting the postion.

Reply:
Good point.